smartsim/_core/control/controller.py

```diff
@@ -212,12 +217,22 @@ def stop_db(self, db: Orchestrator) -> None:
         if db.batch:
             self.stop_entity(db)
         else:
-            shutdown_db(db.hosts, db.ports)
-            with JM_LOCK:
-                for entity in db:
-                    job = self._jobs[entity.name]
-                    job.set_status(STATUS_CANCELLED, "", 0, output=None, error=None)
-                    self._jobs.move_to_completed(job)
+            for node in db.entities:
+                for host_ip, port in itertools.product(
+                    (get_ip_from_host(host) for host in node.hosts), db.ports
+                ):
+                    retcode, _, _ = shutdown_db_node(host_ip, port)
+                    # Sometimes the DB will not shutdown (unless we force NOSAVE)
+                    if retcode != 0:
+                        self.stop_entity(node)
+                        continue
+
+                    with JM_LOCK:
```

Comment:
The problem is that, if we fail sending the shutdown command, we are calling `self.stop_entity(node)` which acquires the lock in the nested scope.

Reply:
Is `JM_LOCK` a `threading.RLock`?

Iirc (and someone should check me on this!) the difference between it and a traditional `threading.Lock` was that while `Lock` acted like a traditional lock, which is "un-ownd" by the thread, a thread is temporarily given "ownership" of an `RLock` so that it could be acquired multiple times by the same thread and is only truly released when the number of `aquire` calls done by a thread equaled the number of `release`s

Reply:
I see, I'll investigate this further, thanks for pointing it out

Reply:
It should then work now!